libcudacxx/include/cuda/std/__complex/logarithms.h

```diff
@@ -23,28 +23,277 @@
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * -8.33332949182035986890e-2);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * 9.09088791817835584208e-2);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * -9.99999883706452485921e-2);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * 0.11111111583539913517);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * -0.12500000038299266536);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * 0.14285714280067426940);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * -0.16666666666197160751);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * 0.20000000000032358560);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * -0.25000000000001904032);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * 0.33333333333333270421);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * -0.5);
+  __log1p_poly        = _CUDA_VSTD::fma(__log1p_poly, __x, 0.5 * 1.0);
+  __log1p_poly *= __x;
+
+  return __log1p_poly;
+}
+
 template <class _Tp>
 [[nodiscard]] _CCCL_API inline complex<_Tp> log(const complex<_Tp>& __x)
 {
-  return complex<_Tp>(_CUDA_VSTD::log(_CUDA_VSTD::abs(__x)), _CUDA_VSTD::arg(__x));
+  // Uint of the same size as our fp type.
+  // Shouldn't need make_unsigned, but just in case:
+  using _UINT_T   = make_unsigned_t<__fp_storage_of_t<_Tp>>;
+  using _CONST_TP = const _Tp;
+
+  // Some needed constants:
+  constexpr _UINT_T __mant_mask = __fp_mant_mask_of_v<_Tp>;
+  constexpr _UINT_T __exp_mask  = __fp_exp_mask_of_v<_Tp>;
+
+  constexpr int32_t __mant_nbits = __fp_mant_nbits_v<__fp_format_of_v<_Tp>>;
+  constexpr int32_t __exp_bias   = __fp_exp_bias_v<__fp_format_of_v<_Tp>>;
+
+  // Cut off the hi and low bit of __exp_mask.
+  // 0x3F000000 for fp32,
+  // 0x3FE0000000000000 for fp64 etc
+  constexpr _UINT_T __exp_mask_of_half = ((__exp_mask >> (__mant_nbits + 2)) << (__mant_nbits + 1));
+
+  _CONST_TP __real_abs = _CUDA_VSTD::fabs(__x.real());
+  _CONST_TP __imag_abs = _CUDA_VSTD::fabs(__x.imag());
+
+  _Tp __max = (__real_abs > __imag_abs) ? __real_abs : __imag_abs;
+  _Tp __min = (__real_abs > __imag_abs) ? __imag_abs : __real_abs;
+
+  // We would like to range reduce these values so that abs(x) ~ 1, as we'll take the log of this.
+  // The below code inlines and removes these two calls:
+  //    _Tp __max_reduced = _CUDA_VSTD::frexpf(__max, &__exp);
+  //    _Tp __min_reduced = _CUDA_VSTD::ldexpf(__min, -__exp);
+
+  // Unbiased exponent of 2.0*__max
+  int32_t __exp = static_cast<int32_t>(reinterpret_cast<_UINT_T&>(__max) >> __mant_nbits) - __exp_bias + 1;
+
+  // Quick frexp:
+  _UINT_T __max_reduced_as_uint = (reinterpret_cast<_UINT_T&>(__max) & __mant_mask) | __exp_mask_of_half;
+  _Tp __max_reduced             = reinterpret_cast<_Tp&>(__max_reduced_as_uint);
```

Comment:
We should rather use `_CUDA_VSTD::bit_cast<__storage_t>(__max)`

That will allow us to actually make this function `constexpr` for most of the supported compilers